checks/raw/license.go

```diff
@@ -69,16 +69,49 @@ func License(c *checker.CheckRequest) (checker.LicenseData, error) {
 	var results checker.LicenseData
 	var path string
 
+	licensesFound, lerr := c.RepoClient.ListLicenses()
+	if lerr == nil && len(licensesFound) > 0 {
+		// TODO: fmt.Printf ("'%T' of size '%d' has '%+v'\n", licensesFound, len(licensesFound), licensesFound)
+		for _, v := range licensesFound {
+			results.LicenseFiles = append(results.LicenseFiles,
+				checker.LicenseFile{
+					File: checker.File{
+						Path: v.Path,
+						Type: checker.FileTypeSource,
+					},
+					License: checker.License{
+						Key:         v.Key,
+						Name:        v.Name,
+						Size:        v.Size,
+						SpdxID:      v.SPDXId,
+						Attribution: checker.LicenseAttributionTypeRepo,
+					},
+				})
+		}
+		return results, nil
+	}
+
+	// no licenses reported by repo API, continue looking for files
 	err := fileparser.OnAllFilesDo(c.RepoClient, isLicenseFile, &path)
 	if err != nil {
 		return results, fmt.Errorf("fileparser.OnAllFilesDo: %w", err)
 	}
 
+	// scorecard search stops at first candidate (isLicenseFile) license file found
 	if path != "" {
```

Comment:
do we need change the score calculation? I've not seen a change, but I may have missed it.
Please also add unit tests to verify the change of behavior

Reply:
I believe the whole point of issue #1369 was to change the scoring. In this PR description above and my note to the reviewers is that I'd like to have a discussion about scoring (knowing now we can get more definitive answers, e.g., confidence, from GH/repo APIs and perhaps that scoring could be predicated on the "source" of the information.

My goal for this change to the license check is motivated by my need... I need (from scorecard, if I'm going to use it) more information about the license actually attached to a project. right now I have to go off and look to see just what that license is if scorecard say's "License Found". Since scorecard is "there" in the moment, just take a few and tell me which one it is.

For me, license check, would be more "usable" by using scorecard to retrieve the specific license detected as it would a) make this information more readily available to the end-user (me); and b) would support develop pipelines that need to sense the license (by name) as a risk management activity without having to move to other means to acquire such license information

I think we can make progress on the changes to the license check to use the repo APIs when possible (and still have the legacy), get that additional meta-data and report it (raw, default, json), and then change the scoring at a later date. As I change the code, I was **NOT** making the assumption that a change to the scoring evaluation would dictate merging this change--but I am happy to change the score evaluation if we can come to a decision.

Reply:
SGTM. I asked about the scoring because the issue you linked in the description mentioned a scoring change. I'm totally onboard updating the scoring later

Reply:
Like Elmer Fudd chasing that “wascally wabbit”—I went down the _wabbit hole_ on the SC legacy code when I tried to update `checks.md` to match this PR.

In doing that review, I discovered a bunch of stuff which tells me that `checks.md` and the code simply don’t match.

I think the answer to **gradated scoring** is right in `checks.md` if the code actually **did** what `checks.md` seems to specify (and GH API is very close to doing this already, except looking inside a top-level dir `LICENSE/`).

**FIrst: scoring approach**

Naming Requirements (one or more) (6/10):
- A license file is in the top-level directory with any combination of the following names: LICENSE, LICENCE, COPYING, COPYRIGHT (regardless of extension)
> for GH API this would be limited to a file named `LICENSE`

- A folder named LICENSE, LICENCE, LICENSES or LICENCES and specific license(s) are named as files within that folder following the SPDX license identifier followed by an appropriate file extension, as described in the REUSE Specification.
> for GH API the SPDX identifier would be deduced from the `LICENSE` file content and be limited to the 13 supported by the GH

Location Requirements (3/10):
- License file is prominent for the repository, that is the license file or folder is in the top-level directory
> for GH API the `LICENSE` file would have to be at the top to work. So basically, if the GH API returns a SPDX identifier the repo would get `6+3=9` score at this point.

Identified license(s) (one or more) (1/10):
- is FSF (see “FSF Free/Libre?” column at https://spdx.org/licenses/)
- is OSI (see “OSI Approved?” column at https://spdx.org/licenses/)
> this can be done, but I have question about how to best implement that sensing. I elaborate on that closer to the bottom of this comment.

Remediation Steps

(pretty much stays the way it is now in `checks.md` without being specific about the extension.

As for FSF and OSI licenses and the list at SPDX.org
- 63 licenses are "recognized" (if that is the right way to say that) by both those organizations (FSF & OSI)
- 64 licenses are OSI only
- 46 licenses are FSF only
> for a total of 173 licenses that are eligible for the +1 up in the evaluation scoring

that would be some massive regex or large set of regexs to compile - but I need help, I have an idea which I think would work:

I took the table at `spdx.org/licenses` and converted it directly into a single json object (well, file, in my case for testing)

Assuming (legacy) found `LICENSE/MulanPSL-2.0.md`

- this would get `6 points` for being there (i.e., `LICENSE/` — a folder) and having a file matching an SPDX identifier (i.e., `MulanPSL-2.0`)
- this would get `3 more points` for being in a prominent location (top-level), e.g., `LICENSE/`
- this would get `1 more point` for being an OSI approved license, by removing the extension (**could be a bit tricky**) a json lookup would show this:
> GH API would not be able to get this extra `+1` has it would not find `MulanPSL-2.0` and the GH API will `404`
```
jq -r '.[] | select(."Identifier"=="MulanPSL-2.0") ' SPDX-license-IDs-202211131030CST.json
{
  "FSF Free/Libre?": "",
  "Full Name": "Mulan Permissive Software License, Version 2",
  "Identifier": "MulanPSL-2.0",
  "OSI Approved?": "Y"
}
```
an associative array/map is the way to go there. The map would have the 173 licences from the SPDX list (which has 466 SPDX ids in it—i’d eliminate any non-FSF/non-OSI licenses)—I'd would like to know if there would be a better way if this approach is even fathomable for SC.

Another example: simply `LICENSE`

- this would get `6 points` for being there (i.e., `LICENSE`)
- this would get `3 more points` for being in a prominent location (top-level), e.g., `LICENSE`
> I would stop there - no extra “point” for FSF/OSI, why?
-- 1. since LICENSE is a file, the SPDX identifer would have to be contained _within_ the file, hence SC have to read the contents and “regex”/“string search” the contents looking for a hit to one of the 173 identifiers
-- 2. if the “community” wanted the filename to be (for example) MulanPSL-2.0.md at the top level (not in a LICENSE/ folder), we’re back to regex searching file names in the repo for one of those 173 identifiers. this is a (n files x 173 liceences to regex search)
> GH API would, however, provide the info needed to assess the final `+1` points

Another example: `.absolutelynothingtoseehere/license` (I actually tried this one with current SC legacy code—I got 10 point score!)

- this would get `6 points` for being there (.i.e., `license`)
- this would NOT get `3 points`, the location is not at the top-level (it is obscured)
- this would NOT get `1 point` for FSF/OSI - same reason as before just too many regex’s and files to search across those regex's

Reply:
Scoring you propose seems reasonable

Reply:
Done.